rpcperms: set CustomCaveatCondition on middleware req

sangaman opened the pull request:
macaroons: add GetCustomCaveatCondition func 

This adds a `GetCustomCaveatCondition` function that returns the custom
caveat condition for a given macaroon and caveat name. Previously there
was no function for getting the custom caveat condition from a macaroon,
only for setting one.

rpcperms: set CustomCaveatCondition on middleware req 

This sets the `CustomCaveatCondition` value on rpc middleware requests
if one exists. Previously, this value was always blank even if the
macaroon had a value set for its custom caveat condition.

Fixes #6164.

Edit: I have tested this locally with a grpc interceptor service and verified that it fixed the related issue, the `custom_caveat_condition` field has a value as expected (and no value when none is set while baking the macaroon).

guggero commented on macaroons/constraints.go:
```diff
@@ -216,3 +216,25 @@ func HasCustomCaveat(mac *macaroon.Macaroon, customCaveatName string) bool {
+
+// GetCustomCaveatCondition returns the custom caveat condition for the given
+// custom caveat name from the given macaroon.
+func GetCustomCaveatCondition(mac *macaroon.Macaroon, customCaveatName string) (string, error) {
```
nit: line length, here and in several other places of the PR. Make sure you use 8 spaces for the tab character in your IDE.

sangaman replied:
Gotcha, I'm surprised this wasn't caught by `go fmt`?

guggero replied:
We can't enforce that with `go fmt` since in some places we do need longer lines. That's why it's currently only in the contribution checklist: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#contribution-checklist

sangaman replied:
Thanks I'll keep that in mind.

guggero commented on macaroons/constraints.go:
```diff
@@ -216,3 +216,25 @@ func HasCustomCaveat(mac *macaroon.Macaroon, customCaveatName string) bool {
+			if len(caveat.Id) > len(caveatPrefix) {
+				return string(caveat.Id[len(caveatPrefix)+1:]), nil
+			}
+			return "", nil // There is no condition for this custom caveat.
```
nit: comment should be above code (and a newline before a comment).

guggero commented on macaroons/constraints.go:
```diff
@@ -216,3 +216,25 @@ func HasCustomCaveat(mac *macaroon.Macaroon, customCaveatName string) bool {
+	for _, caveat := range mac.Caveats() {
+		if bytes.HasPrefix(caveat.Id, caveatPrefix) {
+			if len(caveat.Id) > len(caveatPrefix) {
+				return string(caveat.Id[len(caveatPrefix)+1:]), nil
```
If we do +1 then we also need to do that in the above condition. Otherwise a macaroon with just a space after the name would panic here.

sangaman replied:
I thought this would be fine - if `len(caveat.Id)` is greater than `len(caveatPrefix)` then we should be able to reference `len(caveatPrefix) + 1` within `caveat.Id`. The line above is equivalent to `if len(caveat.Id) >= len(caveatPrefix) + 1` - in fact I'll just change it to that so it's a bit more clear. But I tested the code in its current state and the tests pass fine with a blank space for the caveat condition like `constraintFunc := macaroons.CustomConstraint("unit-test", " ")`.

Roasbeef replied:
Could use a comment here outlining the expected format to aide a reader. 

How about something like this to make it more readable? 
```go
caveatSplit := strings.SplitN(caveat.Id, caveatPrefix, 2)
if len(caveatSplit) == 2 {
    return caveatSplit[1]
}
```

This'll split the string into two parts, one of everything before the prefix, and one of everything after. IIUC, we just want to return everything after.

Roasbeef replied:
Assuming we don't want the whitespace, then we'd add a: `strings.TrimSpace(caveatSplit[1]))`

sangaman replied:
I made this change and added a comment. Note that I needed to wrap `caveat.Id` and `caveatPrefix` with `string()` since they are currently `bytes`.

Rather than trim I added a character of whitespace to the prefix that we match/split on. That means we won't match a caveat with the same name but no condition (since it won't have that whitespace), but the end result is the same as we return an empty string in both cases.

guggero commented on docs/release-notes/release-notes-0.14.2.md:
```diff
@@ -87,10 +87,15 @@ connection from the watch-only node.
   WaitingCloseResp from calling `PendingChannels`.
 
+* [CustomCaveatCondition is now properly set]
+  (https://github.com/lightningnetwork/lnd/pull/6185) on
```
This won't render correctly. You need to break inside the brackets, so for example:
```
* [CustomCaveatCondition is now properly
   set](https://github.com/lightningnetwork/lnd/pull/6185) on
```

sangaman replied:
I didn't know the `(` had to be touching the `]`, thanks.

guggero commented on rpcperms/interceptor.go:
```diff
@@ -929,6 +929,14 @@ func (r *InterceptorChain) acceptRequest(requestID uint64,
 		}
 
+		var err error
+		msg.CustomCaveatCondition, err = macaroons.GetCustomCaveatCondition(msg.Macaroon, middleware.customCaveatName)
```
We'll probably want the same in the `interceptResponse` call below.

guggero replied:
Ah, and we should just ignore the error here. Otherwise in read-only mode this will fail if a normal macaroon without custom condition comes in.

sangaman replied:
I was unsure about whether this logic was needed in `interceptResponse` as well - it's still not clear to me which is used when but I'll add this logic there for good measure.

And good catch about ignoring the error, I overlooked the times when we'll be read only but not have a custom caveat and get past the `if` check above. if think I'll just wrap this in a `if hasCustomCaveat` statement

sangaman replied:
Ignoring the error felt a bit cleaner on second thought, less boilerplate code and we just get back an empty string anyway.

guggero reviewed (commented):
Thank you very much for the fix :tada: 

Looks great so far.
Would be nice if we could add a simple check to the integration tests as well.

Here's a diff with my changes to add the check:
[interceptor.diff.txt](https://github.com/lightningnetwork/lnd/files/7912111/interceptor.diff.txt)

sangaman commented:
PR updated with feedback & patch above. Thank you!

guggero commented on macaroons/constraints.go:
```diff
@@ -216,3 +216,28 @@ func HasCustomCaveat(mac *macaroon.Macaroon, customCaveatName string) bool {
+// GetCustomCaveatCondition returns the custom caveat condition for the given
+// custom caveat name from the given macaroon.
+func GetCustomCaveatCondition(mac *macaroon.Macaroon,
+	customCaveatName string) (string, error) {
```
nit: add newline after multi-line function header.

guggero commented on macaroons/constraints_test.go:
```diff
@@ -132,6 +132,10 @@ func TestCustomConstraint(t *testing.T) {
 	require.False(t, macaroons.HasCustomCaveat(testMacaroon, "something"))
 	require.False(t, macaroons.HasCustomCaveat(nil, "foo"))
 
+	customCaveatCondition, err := macaroons.GetCustomCaveatCondition(testMacaroon, "unit-test")
```
nit: line length

sangaman replied:
Somehow missed these even after triple checking visually for line lengths.

guggero reviewed (approved):
Nice, thanks for the quick update! LGTM :tada: 

Integration test failures are due to known flakes.

sangaman commented:
I updated the PR according to the latest feedback. I also dropped the error return from `GetCustomCaveatCondition` since it wasn't being used anywhere, it will just return empty strings in case the macaroon is nil or there's no match on the caveat name.

Roasbeef commented on macaroons/constraints.go:
```diff
@@ -216,3 +216,37 @@ func HasCustomCaveat(mac *macaroon.Macaroon, customCaveatName string) bool {
+		// "lnd-custom [custom-caveat-name] [custom-caveat-condition]"
+		// and we only want the condition part. If we match the prefix
+		// part we return the condition that comes after the prefix.
+		if bytes.HasPrefix(caveat.Id, caveatPrefix) {
```
👍

Roasbeef reviewed (approved):
LGTM 🏄‍♂️
